Bottom: 06b045eaee28f68625e1725b496d62d3f1888bf6
Top:    291c417d2a6d579ef1c6372e92e9843817748c35
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-12-25 12:40:46 +0000

Refactor signature of PlowsPolygon callback to take userdata

This allows callers to pass arbitrary data into their callbacks
without resorting to nasty workarounds like global variables.


---

diff --git a/src/find.c b/src/find.c
index bea37c2..b5f7fdf 100644
--- a/src/find.c
+++ b/src/find.c
@@ -3485,7 +3485,7 @@ DRCFind (int What, void *ptr1, void *ptr2, void *ptr3)
 
 static int
 drc_callback (DataType *data, LayerType *layer, PolygonType *polygon,
-              int type, void *ptr1, void *ptr2)
+              int type, void *ptr1, void *ptr2, void *userdata)
 {
   char *message;
   Coord x, y;
@@ -3677,7 +3677,7 @@ DRCAll (void)
       COPPERLINE_LOOP (PCB->Data);
       {
         /* check line clearances in polygons */
-        if (PlowsPolygon (PCB->Data, LINE_TYPE, layer, line, drc_callback))
+        if (PlowsPolygon (PCB->Data, LINE_TYPE, layer, line, drc_callback, NULL))
           {
             IsBad = true;
             break;
@@ -3721,7 +3721,7 @@ DRCAll (void)
     {
       COPPERARC_LOOP (PCB->Data);
       {
-        if (PlowsPolygon (PCB->Data, ARC_TYPE, layer, arc, drc_callback))
+        if (PlowsPolygon (PCB->Data, ARC_TYPE, layer, arc, drc_callback, NULL))
           {
             IsBad = true;
             break;
@@ -3765,7 +3765,7 @@ DRCAll (void)
     {
       ALLPIN_LOOP (PCB->Data);
       {
-        if (PlowsPolygon (PCB->Data, PIN_TYPE, element, pin, drc_callback))
+        if (PlowsPolygon (PCB->Data, PIN_TYPE, element, pin, drc_callback, NULL))
           {
             IsBad = true;
             break;
@@ -3841,7 +3841,7 @@ DRCAll (void)
     {
       ALLPAD_LOOP (PCB->Data);
       {
-        if (PlowsPolygon (PCB->Data, PAD_TYPE, element, pad, drc_callback))
+        if (PlowsPolygon (PCB->Data, PAD_TYPE, element, pad, drc_callback, NULL))
           {
             IsBad = true;
             break;
@@ -3885,7 +3885,7 @@ DRCAll (void)
     {
       VIA_LOOP (PCB->Data);
       {
-        if (PlowsPolygon (PCB->Data, VIA_TYPE, via, via, drc_callback))
+        if (PlowsPolygon (PCB->Data, VIA_TYPE, via, via, drc_callback, NULL))
           {
             IsBad = true;
             break;
diff --git a/src/polygon.c b/src/polygon.c
index 2d3e9e4..5f17fd0 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -1458,12 +1458,13 @@ struct plow_info
   LayerType *layer;
   DataType *data;
   int (*callback) (DataType *, LayerType *, PolygonType *, int, void *,
-                   void *);
+                   void *, void *);
+  void *userdata;
 };
 
 static int
 subtract_plow (DataType *Data, LayerType *Layer, PolygonType *Polygon,
-               int type, void *ptr1, void *ptr2)
+               int type, void *ptr1, void *ptr2, void *userdata)
 {
   if (!Polygon->Clipped)
     return 0;
@@ -1496,7 +1497,7 @@ subtract_plow (DataType *Data, LayerType *Layer, PolygonType *Polygon,
 
 static int
 add_plow (DataType *Data, LayerType *Layer, PolygonType *Polygon,
-          int type, void *ptr1, void *ptr2)
+          int type, void *ptr1, void *ptr2, void *userdata)
 {
   switch (type)
     {
@@ -1528,7 +1529,7 @@ plow_callback (const BoxType * b, void *cl)
 
   if (TEST_FLAG (CLEARPOLYFLAG, polygon))
     return plow->callback (plow->data, plow->layer, polygon, plow->type,
-                           plow->ptr1, plow->ptr2);
+                           plow->ptr1, plow->ptr2, plow->userdata);
   return 0;
 }
 
@@ -1536,7 +1537,8 @@ int
 PlowsPolygon (DataType * Data, int type, void *ptr1, void *ptr2,
               int (*call_back) (DataType *data, LayerType *lay,
                                 PolygonType *poly, int type, void *ptr1,
-                                void *ptr2))
+                                void *ptr2, void *userdata),
+              void *userdata)
 {
   BoxType sb = ((PinType *) ptr2)->BoundingBox;
   int r = 0;
@@ -1547,6 +1549,7 @@ PlowsPolygon (DataType * Data, int type, void *ptr1, void *ptr2,
   info.ptr2 = ptr2;
   info.data = Data;
   info.callback = call_back;
+  info.userdata = userdata;
   switch (type)
     {
     case PIN_TYPE:
@@ -1609,12 +1612,12 @@ PlowsPolygon (DataType * Data, int type, void *ptr1, void *ptr2,
       {
         PIN_LOOP ((ElementType *) ptr1);
         {
-          PlowsPolygon (Data, PIN_TYPE, ptr1, pin, call_back);
+          PlowsPolygon (Data, PIN_TYPE, ptr1, pin, call_back, userdata);
         }
         END_LOOP;
         PAD_LOOP ((ElementType *) ptr1);
         {
-          PlowsPolygon (Data, PAD_TYPE, ptr1, pad, call_back);
+          PlowsPolygon (Data, PAD_TYPE, ptr1, pad, call_back, userdata);
         }
         END_LOOP;
       }
@@ -1632,7 +1635,7 @@ RestoreToPolygon (DataType * Data, int type, void *ptr1, void *ptr2)
   if (type == POLYGON_TYPE)
     InitClip (PCB->Data, (LayerType *) ptr1, (PolygonType *) ptr2);
   else
-    PlowsPolygon (Data, type, ptr1, ptr2, add_plow);
+    PlowsPolygon (Data, type, ptr1, ptr2, add_plow, NULL);
 }
 
 void
@@ -1644,7 +1647,7 @@ ClearFromPolygon (DataType * Data, int type, void *ptr1, void *ptr2)
   if (type == POLYGON_TYPE)
     InitClip (PCB->Data, (LayerType *) ptr1, (PolygonType *) ptr2);
   else
-    PlowsPolygon (Data, type, ptr1, ptr2, subtract_plow);
+    PlowsPolygon (Data, type, ptr1, ptr2, subtract_plow, NULL);
 }
 
 bool
diff --git a/src/polygon.h b/src/polygon.h
index c217653..64d9a36 100644
--- a/src/polygon.h
+++ b/src/polygon.h
@@ -63,7 +63,8 @@ int PolygonHoles (PolygonType *ptr, const BoxType *range,
 		  int (*callback) (PLINE *, void *user_data),
                   void *user_data);
 int PlowsPolygon (DataType *, int, void *, void *,
-		  int (*callback) (DataType *, LayerType *, PolygonType *, int, void *, void *));
+                  int (*callback) (DataType *, LayerType *, PolygonType *, int, void *, void *, void *),
+                  void *userdata);
 void ComputeNoHoles (PolygonType *poly);
 POLYAREA * ContourToPoly (PLINE *);
 POLYAREA * PolygonToPoly (PolygonType *);
